File: tests/rosetta/transpiler/php/arithmetic-evaluation.php
<?php
ini_set('memory_limit', '-1');
function skipWS($p) {
  global $parseIntStr, $parseNumber, $parseFactor, $powInt, $parsePower, $parseTerm, $parseExpr, $evalExpr, $main;
  $i = $p['pos'];
  while ($i < strlen($p['expr']) && substr($p['expr'], $i, $i + 1 - $i) == ' ') {
  $i = $i + 1;
};
  $p['pos'] = $i;
  return $p;
}
function parseIntStr($str) {
  global $skipWS, $parseNumber, $parseFactor, $powInt, $parsePower, $parseTerm, $parseExpr, $evalExpr, $main;
  $i = 0;
  $n = 0;
  while ($i < strlen($str)) {
  $n = $n * 10 + (ord(substr($str, $i, $i + 1 - $i))) - 48;
  $i = $i + 1;
};
  return $n;
}
function parseNumber($p) {
  global $skipWS, $parseIntStr, $parseFactor, $powInt, $parsePower, $parseTerm, $parseExpr, $evalExpr, $main;
  $p = skipWS($p);
  $start = $p['pos'];
  while ($p['pos'] < strlen($p['expr'])) {
  $ch = substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']);
  if ($ch >= '0' && $ch <= '9') {
  $p['pos'] = $p['pos'] + 1;
} else {
  break;
}
};
  $token = substr($p['expr'], $start, $p['pos'] - $start);
  return ['v' => parseIntStr($token), 'p' => $p];
}
function parseFactor($p) {
  global $skipWS, $parseIntStr, $parseNumber, $powInt, $parsePower, $parseTerm, $parseExpr, $evalExpr, $main;
  $p = skipWS($p);
  if ($p['pos'] < strlen($p['expr']) && substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']) == '(') {
  $p['pos'] = $p['pos'] + 1;
  $r = parseExpr($p);
  $v = $r['v'];
  $p = $r['p'];
  $p = skipWS($p);
  if ($p['pos'] < strlen($p['expr']) && substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']) == ')') {
  $p['pos'] = $p['pos'] + 1;
};
  return ['v' => $v, 'p' => $p];
}
  if ($p['pos'] < strlen($p['expr']) && substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']) == '-') {
  $p['pos'] = $p['pos'] + 1;
  $r = parseFactor($p);
  $v = $r['v'];
  $p = $r['p'];
  return ['v' => -$v, 'p' => $p];
}
  return parseNumber($p);
}
function powInt($base, $exp) {
  global $skipWS, $parseIntStr, $parseNumber, $parseFactor, $parsePower, $parseTerm, $parseExpr, $evalExpr, $main;
  $r = 1;
  $b = $base;
  $e = $exp;
  while ($e > 0) {
  if ($e % 2 == 1) {
  $r = $r * $b;
}
  $b = $b * $b;
  $e = $e / intval(2);
};
  return $r;
}
function parsePower($p) {
  global $skipWS, $parseIntStr, $parseNumber, $parseFactor, $powInt, $parseTerm, $parseExpr, $evalExpr, $main;
  $r = parseFactor($p);
  $v = $r['v'];
  $p = $r['p'];
  while (true) {
  $p = skipWS($p);
  if ($p['pos'] < strlen($p['expr']) && substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']) == '^') {
  $p['pos'] = $p['pos'] + 1;
  $r2 = parseFactor($p);
  $rhs = $r2['v'];
  $p = $r2['p'];
  $v = powInt($v, $rhs);
} else {
  break;
}
};
  return ['v' => $v, 'p' => $p];
}
function parseTerm($p) {
  global $skipWS, $parseIntStr, $parseNumber, $parseFactor, $powInt, $parsePower, $parseExpr, $evalExpr, $main;
  $r = parsePower($p);
  $v = $r['v'];
  $p = $r['p'];
  while (true) {
  $p = skipWS($p);
  if ($p['pos'] < strlen($p['expr'])) {
  $op = substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']);
  if ($op == '*') {
  $p['pos'] = $p['pos'] + 1;
  $r2 = parsePower($p);
  $rhs = $r2['v'];
  $p = $r2['p'];
  $v = $v * $rhs;
  continue;
};
  if ($op == '/') {
  $p['pos'] = $p['pos'] + 1;
  $r2 = parsePower($p);
  $rhs = $r2['v'];
  $p = $r2['p'];
  $v = $v / intval($rhs);
  continue;
};
}
  break;
};
  return ['v' => $v, 'p' => $p];
}
function parseExpr($p) {
  global $skipWS, $parseIntStr, $parseNumber, $parseFactor, $powInt, $parsePower, $parseTerm, $evalExpr, $main;
  $r = parseTerm($p);
  $v = $r['v'];
  $p = $r['p'];
  while (true) {
  $p = skipWS($p);
  if ($p['pos'] < strlen($p['expr'])) {
  $op = substr($p['expr'], $p['pos'], $p['pos'] + 1 - $p['pos']);
  if ($op == '+') {
  $p['pos'] = $p['pos'] + 1;
  $r2 = parseTerm($p);
  $rhs = $r2['v'];
  $p = $r2['p'];
  $v = $v + $rhs;
  continue;
};
  if ($op == '-') {
  $p['pos'] = $p['pos'] + 1;
  $r2 = parseTerm($p);
  $rhs = $r2['v'];
  $p = $r2['p'];
  $v = $v - $rhs;
  continue;
};
}
  break;
};
  return ['v' => $v, 'p' => $p];
}
function evalExpr($expr) {
  global $skipWS, $parseIntStr, $parseNumber, $parseFactor, $powInt, $parsePower, $parseTerm, $parseExpr, $main;
  $p = ['expr' => $expr, 'pos' => 0];
  $r = parseExpr($p);
  return $r['v'];
}
function main() {
  global $skipWS, $parseIntStr, $parseNumber, $parseFactor, $powInt, $parsePower, $parseTerm, $parseExpr, $evalExpr;
  $expr = '2*(3-1)+2*5';
  echo $expr . ' = ' . json_encode(evalExpr($expr), 1344), PHP_EOL;
}
main();

File: tests/rosetta/transpiler/FS/arithmetic-evaluation.fs
// Generated 2025-07-25 00:53 +0700

exception Break
exception Continue

exception Return

type Parser = {
    expr: string
    pos: int
}
type Res = {
    v: int
    p: Parser
}
let rec skipWS (p: Parser) =
    let mutable __ret : Parser = Unchecked.defaultof<Parser>
    let mutable p = p
    try
        let mutable i: int = p.pos
        while (i < (String.length (p.expr))) && ((p.expr.Substring(i, (i + 1) - i)) = " ") do
            i <- i + 1
        { p with pos = i }
        __ret <- p
        raise Return
        __ret
    with
        | Return -> __ret
and parseIntStr (str: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable str = str
    try
        let mutable i: int = 0
        let mutable n: int = 0
        while i < (String.length str) do
            n <- ((n * 10) + (int (str.Substring(i, (i + 1) - i)))) - 48
            i <- i + 1
        __ret <- n
        raise Return
        __ret
    with
        | Return -> __ret
and parseNumber (p: Parser) =
    let mutable __ret : Res = Unchecked.defaultof<Res>
    let mutable p = p
    try
        p <- skipWS p
        let mutable start: int = p.pos
        try
            while (p.pos) < (String.length (p.expr)) do
                let ch: string = p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))
                if (ch >= "0") && (ch <= "9") then
                    { p with pos = (p.pos) + 1 }
                else
                    raise Break
        with
        | Break -> ()
        | Continue -> ()
        let token: string = p.expr.Substring(start, (p.pos) - start)
        __ret <- { v = parseIntStr token; p = p }
        raise Return
        __ret
    with
        | Return -> __ret
and parseFactor (p: Parser) =
    let mutable __ret : Res = Unchecked.defaultof<Res>
    let mutable p = p
    try
        p <- skipWS p
        if ((p.pos) < (String.length (p.expr))) && ((p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))) = "(") then
            { p with pos = (p.pos) + 1 }
            let mutable r = parseExpr p
            let mutable v = r.v
            p <- r.p
            p <- skipWS p
            if ((p.pos) < (String.length (p.expr))) && ((p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))) = ")") then
                { p with pos = (p.pos) + 1 }
            __ret <- { v = v; p = p }
            raise Return
        if ((p.pos) < (String.length (p.expr))) && ((p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))) = "-") then
            { p with pos = (p.pos) + 1 }
            let mutable r = parseFactor p
            let mutable v = r.v
            p <- r.p
            __ret <- { v = -v; p = p }
            raise Return
        __ret <- parseNumber p
        raise Return
        __ret
    with
        | Return -> __ret
and powInt (``base``: int) (exp: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable r: int = 1
        let mutable b: int = ``base``
        let mutable e: int = exp
        while e > 0 do
            if (e % 2) = 1 then
                r <- r * b
            b <- b * b
            e <- e / (int 2)
        __ret <- r
        raise Return
        __ret
    with
        | Return -> __ret
and parsePower (p: Parser) =
    let mutable __ret : Res = Unchecked.defaultof<Res>
    let mutable p = p
    try
        let mutable r = parseFactor p
        let mutable v = r.v
        p <- r.p
        try
            while true do
                p <- skipWS p
                if ((p.pos) < (String.length (p.expr))) && ((p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))) = "^") then
                    { p with pos = (p.pos) + 1 }
                    let mutable r2 = parseFactor p
                    let mutable rhs = r2.v
                    p <- r2.p
                    v <- powInt v rhs
                else
                    raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- { v = v; p = p }
        raise Return
        __ret
    with
        | Return -> __ret
and parseTerm (p: Parser) =
    let mutable __ret : Res = Unchecked.defaultof<Res>
    let mutable p = p
    try
        let mutable r = parsePower p
        let mutable v = r.v
        p <- r.p
        try
            while true do
                p <- skipWS p
                if (p.pos) < (String.length (p.expr)) then
                    let op: string = p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))
                    if op = "*" then
                        { p with pos = (p.pos) + 1 }
                        let mutable r2 = parsePower p
                        let mutable rhs = r2.v
                        p <- r2.p
                        v <- v * rhs
                        raise Continue
                    if op = "/" then
                        { p with pos = (p.pos) + 1 }
                        let mutable r2 = parsePower p
                        let mutable rhs = r2.v
                        p <- r2.p
                        v <- v / (int rhs)
                        raise Continue
                raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- { v = v; p = p }
        raise Return
        __ret
    with
        | Return -> __ret
and parseExpr (p: Parser) =
    let mutable __ret : Res = Unchecked.defaultof<Res>
    let mutable p = p
    try
        let mutable r = parseTerm p
        let mutable v = r.v
        p <- r.p
        try
            while true do
                p <- skipWS p
                if (p.pos) < (String.length (p.expr)) then
                    let op: string = p.expr.Substring(p.pos, ((p.pos) + 1) - (p.pos))
                    if op = "+" then
                        { p with pos = (p.pos) + 1 }
                        let mutable r2 = parseTerm p
                        let mutable rhs = r2.v
                        p <- r2.p
                        v <- v + rhs
                        raise Continue
                    if op = "-" then
                        { p with pos = (p.pos) + 1 }
                        let mutable r2 = parseTerm p
                        let mutable rhs = r2.v
                        p <- r2.p
                        v <- v - rhs
                        raise Continue
                raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- { v = v; p = p }
        raise Return
        __ret
    with
        | Return -> __ret
and evalExpr (expr: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable expr = expr
    try
        let mutable p: Parser = { expr = expr; pos = 0 }
        let r = parseExpr p
        __ret <- r.v
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let expr: string = "2*(3-1)+2*5"
        printfn "%s" ((expr + " = ") + (string (evalExpr expr)))
        __ret
    with
        | Return -> __ret
main()
